File: src/python/modelmesh/interfaces/provider.py
"""Provider connector interface and associated data types.

Defines the abstract ProviderConnector interface and all request/response
data types used for OpenAI-compatible model execution. Provider connectors
bridge the library's capability-driven routing model with concrete provider
APIs, handling authentication, format translation, streaming, usage tracking,
and error classification.
"""
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import AsyncIterator, Optional


@dataclass
class ModelPricing:
    """Per-token and per-request cost metadata for a model."""

    input_per_1k_tokens: float = 0.0
    output_per_1k_tokens: float = 0.0
    per_request: float = 0.0


@dataclass
class ModelInfo:
    """Describes a model exposed by a provider connector.

    Model IDs use dot-notation (e.g. ``"openai.gpt-4o"``) to namespace
    by provider. Capabilities, features, and delivery modes feed pool
    membership and routing decisions.
    """

    id: str
    name: str
    capabilities: list[str] = field(default_factory=list)
    context_window: int = 0
    max_output_tokens: int = 0
    pricing: Optional[ModelPricing] = None
    features: dict[str, bool] = field(default_factory=dict)
    delivery: dict[str, bool] = field(
        default_factory=lambda: {"synchronous": True}
    )


@dataclass
class TokenUsage:
    """Token consumption for a single completion request."""

    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


@dataclass
class ChatMessage:
    """A single message in a chat conversation.

    Used both for full messages (in non-streaming responses) and for
    partial deltas (in streaming responses).
    """

    role: str = "assistant"
    content: Optional[str] = None
    tool_calls: Optional[list] = None


@dataclass
class CompletionChoice:
    """One candidate completion returned by the model.

    For streaming responses, ``delta`` carries the incremental content
    and ``message`` is ``None``. For non-streaming responses, ``message``
    carries the full content and ``delta`` is ``None``.
    """

    index: int = 0
    message: Optional[ChatMessage] = None
    delta: Optional[ChatMessage] = None
    finish_reason: Optional[str] = None


@dataclass
class CompletionRequest:
    """OpenAI ChatCompletion-compatible request payload.

    Passed to :meth:`ProviderConnector.complete` and
    :meth:`ProviderConnector.stream`. The ``model`` field uses
    dot-notated IDs (e.g. ``"openai.gpt-4o"``).
    """

    model: str
    messages: list[dict]
    temperature: float = 1.0
    max_tokens: Optional[int] = None
    stream: bool = False
    tools: Optional[list] = None
    top_p: float = 1.0
    stop: Optional[list[str]] = None


@dataclass
class CompletionResponse:
    """OpenAI ChatCompletion-compatible response.

    Attribute access matches the OpenAI SDK shape so callers can use
    ``response.choices[0].message.content`` without adaptation.
    """

    id: str = ""
    model: str = ""
    choices: list[CompletionChoice] = field(default_factory=list)
    usage: TokenUsage = field(default_factory=TokenUsage)
    created: int = 0
    object: str = "chat.completion"


@dataclass
class QuotaStatus:
    """Current quota consumption for a provider."""

    used: int = 0
    limit: Optional[int] = None
    remaining: Optional[int] = None
    reset_at: Optional[str] = None


@dataclass
class RateLimitStatus:
    """Current rate-limit headroom for a provider."""

    requests_remaining: Optional[int] = None
    tokens_remaining: Optional[int] = None
    reset_at: Optional[str] = None


@dataclass
class ErrorClassification:
    """Structured classification of a provider error.

    Used by the router to decide between retry, rotation, or surfacing
    the error to the caller.
    """

    retryable: bool = False
    error_code: Optional[int] = None
    message: str = ""
    category: str = "unknown"  # "auth", "rate_limit", "server", "client", "network"


class ProviderConnector(ABC):
    """Abstract interface for provider connectors.

    A provider connector exposes one or more AI models through a uniform,
    OpenAI-compatible API. It translates requests, manages authentication,
    tracks usage, and reports operational data that drives routing and
    rotation decisions.

    Implementations should subclass this and provide concrete logic for
    each abstract method. The :meth:`is_retryable` convenience method
    and :meth:`close` lifecycle hook have default implementations.
    """

    @abstractmethod
    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        """Execute a chat completion request and return the full response."""
        ...

    @abstractmethod
    async def stream(
        self, request: CompletionRequest
    ) -> AsyncIterator[CompletionResponse]:
        """Execute a streaming chat completion, yielding response chunks."""
        ...

    @abstractmethod
    def get_capabilities(self) -> list[str]:
        """Return the list of capabilities this provider supports."""
        ...

    @abstractmethod
    def supports(self, capability: str) -> bool:
        """Check whether this provider supports a specific capability."""
        ...

    @abstractmethod
    def list_models(self) -> list[ModelInfo]:
        """Return all models available through this provider."""
        ...

    @abstractmethod
    def get_model_info(self, model_id: str) -> ModelInfo:
        """Return metadata for a specific model by its dot-notated ID."""
        ...

    @abstractmethod
    def check_quota(self) -> QuotaStatus:
        """Return current quota consumption and remaining capacity."""
        ...

    @abstractmethod
    def get_rate_limits(self) -> RateLimitStatus:
        """Return current rate-limit headroom."""
        ...

    @abstractmethod
    def get_pricing(self, model_id: str) -> ModelPricing:
        """Return pricing metadata for a specific model."""
        ...

    @abstractmethod
    def report_usage(self, model_id: str, usage: TokenUsage) -> None:
        """Record token usage for internal tracking and budget enforcement."""
        ...

    @abstractmethod
    def classify_error(self, error: Exception) -> ErrorClassification:
        """Classify a provider error for retry and rotation decisions."""
        ...

    def is_retryable(self, error: Exception) -> bool:
        """Convenience method: return True if the error is retryable."""
        return self.classify_error(error).retryable

    async def close(self) -> None:
        """Release any resources held by this connector.

        Called during library shutdown. The default implementation is a
        no-op; subclasses may override to close HTTP sessions, flush
        buffers, etc.
        """
        pass


__all__ = [
    "ModelPricing",
    "ModelInfo",
    "TokenUsage",
    "ChatMessage",
    "CompletionChoice",
    "CompletionRequest",
    "CompletionResponse",
    "QuotaStatus",
    "RateLimitStatus",
    "ErrorClassification",
    "ProviderConnector",
]

File: src/typescript/src/interfaces/provider.ts
/**
 * Provider connector interface and associated data types.
 *
 * Defines the ProviderConnector interface and all request/response
 * data types used for OpenAI-compatible model execution.
 */

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

export interface ModelPricing {
  inputPer1kTokens: number;
  outputPer1kTokens: number;
  perRequest: number;
}

export interface ModelInfo {
  id: string;
  name: string;
  capabilities: string[];
  contextWindow: number;
  maxOutputTokens: number;
  pricing?: ModelPricing;
  features: Record<string, boolean>;
  delivery: Record<string, boolean>;
}

export interface TokenUsage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

export interface ChatMessage {
  role: string;
  content?: string;
  toolCalls?: unknown[];
}

export interface CompletionChoice {
  index: number;
  message?: ChatMessage;
  delta?: ChatMessage;
  finishReason?: string;
}

export interface CompletionRequest {
  model: string;
  messages: Record<string, unknown>[];
  temperature: number;
  maxTokens?: number;
  stream: boolean;
  tools?: unknown[];
  topP: number;
  stop?: string[];
}

export interface CompletionResponse {
  id: string;
  model: string;
  choices: CompletionChoice[];
  usage: TokenUsage;
  created: number;
  object: string;
}

export interface QuotaStatus {
  used: number;
  limit?: number;
  remaining?: number;
  resetAt?: string;
}

export interface RateLimitStatus {
  requestsRemaining?: number;
  tokensRemaining?: number;
  resetAt?: string;
}

export interface ErrorClassification {
  retryable: boolean;
  errorCode?: number;
  message: string;
  category: string;
}

// ---------------------------------------------------------------------------
// Audio data types
// ---------------------------------------------------------------------------

/**
 * Request payload for text-to-speech synthesis.
 *
 * Matches the OpenAI audio speech API shape so callers can use
 * `client.audio.speech.create({ model, voice, input })`.
 */
export interface AudioSpeechRequest {
  /** Model ID (e.g. "tts-1", "eleven_multilingual_v2"). */
  model: string;
  /** Text to synthesize. */
  input: string;
  /** Voice identifier (provider-specific). */
  voice: string;
  /** Output audio format. */
  responseFormat?: string;
  /** Speech speed multiplier (0.25–4.0). */
  speed?: number;
}

/**
 * Response from a text-to-speech synthesis request.
 *
 * Wraps audio output metadata. The actual audio bytes are transport-
 * dependent (returned as a Buffer in Node.js, ArrayBuffer in browsers).
 */
export interface AudioSpeechResponse {
  /** Audio content as bytes. null if only metadata is returned. */
  audioData: Uint8Array | null;
  /** Audio format (e.g. "mp3", "wav", "opus"). */
  format: string;
  /** Size of audio data in bytes. */
  sizeBytes: number;
  /** Duration of audio in seconds (if known). */
  durationSeconds?: number;
  /** The model used. */
  model: string;
  /** Input character count. */
  inputCharacters: number;
}

/**
 * Request payload for speech-to-text transcription.
 *
 * Matches the OpenAI audio transcriptions API shape.
 */
export interface AudioTranscriptionRequest {
  /** Model ID (e.g. "whisper-1", "assemblyai-best"). */
  model: string;
  /** Audio file URL or base64-encoded audio data. */
  file: string;
  /** Language hint (ISO-639-1). */
  language?: string;
  /** Output format ("json", "text", "srt", "vtt"). */
  responseFormat?: string;
  /** Prompt to guide transcription. */
  prompt?: string;
}

/**
 * Response from a speech-to-text transcription request.
 */
export interface AudioTranscriptionResponse {
  /** Transcribed text. */
  text: string;
  /** Language detected. */
  language?: string;
  /** Audio duration in seconds. */
  durationSeconds?: number;
  /** Transcription confidence (0.0–1.0). */
  confidence?: number;
  /** The model used. */
  model: string;
}

// ---------------------------------------------------------------------------
// Abstract interface
// ---------------------------------------------------------------------------

export interface ProviderConnector {
  complete(request: CompletionRequest): Promise<CompletionResponse>;
  stream(request: CompletionRequest): AsyncIterableIterator<CompletionResponse>;
  getCapabilities(): string[];
  supports(capability: string): boolean;
  listModels(): ModelInfo[];
  getModelInfo(modelId: string): ModelInfo;
  checkQuota(): QuotaStatus;
  getRateLimits(): RateLimitStatus;
  getPricing(modelId: string): ModelPricing;
  reportUsage(modelId: string, usage: TokenUsage): void;
  classifyError(error: Error): ErrorClassification;
  isRetryable(error: Error): boolean;
  close(): Promise<void>;
}

// ---------------------------------------------------------------------------
// Helper factory functions for creating default data objects
// ---------------------------------------------------------------------------

export function createDefaultTokenUsage(
  overrides?: Partial<TokenUsage>
): TokenUsage {
  return { promptTokens: 0, completionTokens: 0, totalTokens: 0, ...overrides };
}

export function createDefaultCompletionRequest(
  overrides: Partial<CompletionRequest> & { model: string; messages: Record<string, unknown>[] }
): CompletionRequest {
  return {
    temperature: 1.0,
    stream: false,
    topP: 1.0,
    ...overrides,
  };
}

export function createDefaultCompletionResponse(
  overrides?: Partial<CompletionResponse>
): CompletionResponse {
  return {
    id: '',
    model: '',
    choices: [],
    usage: createDefaultTokenUsage(),
    created: 0,
    object: 'chat.completion',
    ...overrides,
  };
}

export function createDefaultModelInfo(
  overrides: Partial<ModelInfo> & { id: string; name: string }
): ModelInfo {
  return {
    capabilities: [],
    contextWindow: 0,
    maxOutputTokens: 0,
    features: {},
    delivery: { synchronous: true },
    ...overrides,
  };
}

export function createDefaultModelPricing(
  overrides?: Partial<ModelPricing>
): ModelPricing {
  return {
    inputPer1kTokens: 0,
    outputPer1kTokens: 0,
    perRequest: 0,
    ...overrides,
  };
}

/**
 * Default implementation of isRetryable -- delegates to classifyError.
 *
 * Matches the Python ProviderConnector.is_retryable() default behavior.
 * Provider implementors can use this as their isRetryable implementation:
 *
 *   isRetryable: (error) => defaultIsRetryable(this, error),
 */
export function defaultIsRetryable(
  connector: Pick<ProviderConnector, 'classifyError'>,
  error: Error
): boolean {
  return connector.classifyError(error).retryable;
}

/**
 * Create a default HealthReport with auto-populated timestamp.
 *
 * Matches the Python HealthReport.__post_init__() behavior where
 * timestamp defaults to the current time if not provided.
 */
export function createDefaultHealthReport(
  overrides: { providerId: string; available: boolean } & Record<string, unknown>
): {
  providerId: string;
  available: boolean;
  latencyMs?: number;
  statusCode?: number;
  error?: string;
  availabilityScore: number;
  timestamp: Date;
} {
  return {
    availabilityScore: 1.0,
    timestamp: new Date(),
    ...overrides,
  } as {
    providerId: string;
    available: boolean;
    latencyMs?: number;
    statusCode?: number;
    error?: string;
    availabilityScore: number;
    timestamp: Date;
  };
}

// ---------------------------------------------------------------------------
// Audio factory functions
// ---------------------------------------------------------------------------

export function createDefaultAudioSpeechRequest(
  overrides: Partial<AudioSpeechRequest> & { model: string; input: string; voice: string }
): AudioSpeechRequest {
  return {
    responseFormat: 'mp3',
    speed: 1.0,
    ...overrides,
  };
}

export function createDefaultAudioSpeechResponse(
  overrides?: Partial<AudioSpeechResponse>
): AudioSpeechResponse {
  return {
    audioData: null,
    format: 'mp3',
    sizeBytes: 0,
    model: '',
    inputCharacters: 0,
    ...overrides,
  };
}

export function createDefaultAudioTranscriptionRequest(
  overrides: Partial<AudioTranscriptionRequest> & { model: string; file: string }
): AudioTranscriptionRequest {
  return {
    responseFormat: 'json',
    ...overrides,
  };
}

export function createDefaultAudioTranscriptionResponse(
  overrides?: Partial<AudioTranscriptionResponse>
): AudioTranscriptionResponse {
  return {
    text: '',
    model: '',
    ...overrides,
  };
}

// Aliases for convenience
export const createModelInfo = createDefaultModelInfo;
export const createModelPricing = createDefaultModelPricing;
export const createTokenUsage = createDefaultTokenUsage;
export const createCompletionResponse = createDefaultCompletionResponse;
